File: 2026-01-01_week04/D/solutions/solution.py
import sys
from collections import defaultdict

input = sys.stdin.readline

n = int(input())
a = list(map(int, input().split()))

cnt = defaultdict(int)
ans = 0

powers = [1 << i for i in range(1, 32)]  # 2^1 تا 2^31

for x in a:
    for p in powers:
        need = p - x
        if need in cnt:
            ans += cnt[need]
    cnt[x] += 1

print(ans)

File: 2026-01-01_week04/D/solutions/solution.java
import java.io.*;
import java.util.*;

public class solution {
    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        
        int n = Integer.parseInt(br.readLine().trim());
        StringTokenizer st = new StringTokenizer(br.readLine());

        HashMap<Long, Long> cnt = new HashMap<>();
        long ans = 0;

        for (int i = 0; i < n; i++) {
            long x = Long.parseLong(st.nextToken());

            for (int p = 1; p <= 31; p++) {
                long power = 1L << p;
                long need = power - x;
                ans += cnt.getOrDefault(need, 0L);
            }

            cnt.put(x, cnt.getOrDefault(x, 0L) + 1);
        }

        System.out.println(ans);
    }
}
